src/parquet/column/statistics.h

```diff
@@ -0,0 +1,128 @@
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#ifndef PARQUET_COLUMN_STATISTICS_H
+#define PARQUET_COLUMN_STATISTICS_H
+
+#include <cstdint>
+#include <string>
+
+#include "parquet/types.h"
+#include "parquet/schema/descriptor.h"
+#include "parquet/util/mem-allocator.h"
+#include "parquet/util/visibility.h"
+
+namespace parquet {
+
+class PARQUET_EXPORT RowGroupStatistics {
+ public:
+  int64_t null_count() const { return null_count_; }
+  int64_t distinct_count() const { return distinct_count_; }
+  int64_t num_values() const { return num_values_; }
+
+  virtual bool HasMinMax() const = 0;
+  virtual void Reset() = 0;
+  virtual void Merge(const RowGroupStatistics& other) = 0;
+
+  // Plain-encoded minimum value
+  virtual std::string EncodedMin() = 0;
+
+  // Plain-encoded maximum value
+  virtual std::string EncodedMax() = 0;
+
+  virtual ~RowGroupStatistics() {}
+
+  Type::type physical_type() const { return descr_->physical_type(); }
+
+ protected:
+  const ColumnDescriptor* descr() const { return descr_; }
+  void SetDescr(const ColumnDescriptor* schema) { descr_ = schema; }
+
+  void IncrementNullCount(int64_t n) { null_count_ += n; }
+
+  void IncrementNumValues(int64_t n) { num_values_ += n; }
+
+  void IncrementDistinctCount(int64_t n) { distinct_count_ += n; }
+
+  void MergeCounts(const RowGroupStatistics& other) {
+    this->null_count_ += other.null_count_;
+    this->distinct_count_ += other.distinct_count_;
+    this->num_values_ += other.num_values_;
+  }
+
+  void ResetCounts() {
+    this->null_count_ = 0;
+    this->distinct_count_ = 0;
+    this->num_values_ = 0;
+  }
+
+  const ColumnDescriptor* descr_ = nullptr;
+  int64_t null_count_ = 0;
+  int64_t distinct_count_ = 0;
+  int64_t num_values_ = 0;
+};
+
+template <typename DType>
+class PARQUET_EXPORT TypedRowGroupStatistics : public RowGroupStatistics {
+ public:
+  using T = typename DType::c_type;
+
+  TypedRowGroupStatistics(
+      const ColumnDescriptor* schema, MemoryAllocator* allocator = default_allocator());
+
+  TypedRowGroupStatistics(const T& min, const T& max, int64_t num_values,
+      int64_t null_count, int64_t distinct_count);
+
+  TypedRowGroupStatistics(const ColumnDescriptor* schema, const std::string& encoded_min,
+      const std::string& encoded_max, int64_t num_values, int64_t null_count,
+      int64_t distinct_count, MemoryAllocator* allocator = default_allocator());
+
+  bool HasMinMax() const override;
+  void Reset() override;
+  void Merge(const RowGroupStatistics& other) override;
+
+  void Update(const T* values, int64_t num_not_null, int64_t num_null);
+
+  const T& min() const { return min_; }
+  const T& max() const { return max_; }
+
+  std::string EncodedMin() override;
+  std::string EncodedMax() override;
+
+ private:
+  MemoryAllocator* allocator_;
+
+  bool has_min_max_ = false;
+  T min_;
+  T max_;
```

Comment:
Yes, thanks for pointing this subtlety out. I'll try to come up with a failing test.